util/extsort/disk_sorter.go

```diff
@@ -15,37 +15,654 @@
+	*h = old[:n-1]
+	return x
+}
+
+type openIterFunc func(file *fileMetadata) (Iterator, error)
+
+// newMergingIter returns an iterator that merges the given files.
+// orderedFiles must be ordered by start key, otherwise it panics.
+func newMergingIter(orderedFiles []*fileMetadata, openIter openIterFunc) *mergingIter {
+	if !slices.IsSortedFunc(orderedFiles, func(a, b *fileMetadata) bool {
+		return bytes.Compare(a.startKey, b.startKey) < 0
+	}) {
+		panic("newMergingIter: orderedFiles are not ordered by start key")
+	}
+	return &mergingIter{
+		orderedFiles: orderedFiles,
+		openIter:     openIter,
+	}
+}
+
+func (m *mergingIter) Seek(key []byte) bool {
+	m.err = nil
+	oldHeap := m.heap
+	m.heap = nil
+
+	var toClose []mergingIterItem
+	openedIters := make(map[int]struct{})
+	for i, item := range oldHeap {
+		file := m.orderedFiles[item.index]
+		if bytes.Compare(key, file.startKey) >= 0 &&
+			bytes.Compare(key, file.endKey) < 0 &&
+			item.Seek(key) {
+			openedIters[item.index] = struct{}{}
+			heap.Push(&m.heap, item)
+		} else {
+			m.err = item.Error()
+			if m.err != nil {
+				// Iterators in m.heap should be closed by mergingIter.Close().
+				toClose = append(toClose, oldHeap[i:]...)
+				break
+			}
+			// The seek key is not in the range of [startKey, endKey), close it.
+			toClose = append(toClose, item)
+		}
+	}
+	for _, item := range toClose {
+		if err := item.Close(); err != nil && m.err == nil {
+			m.err = err
+		}
+	}
+	if m.err != nil {
+		return false
+	}
+
+	m.nextFileIndex = len(m.orderedFiles)
+	for i, file := range m.orderedFiles {
+		if bytes.Compare(file.startKey, key) > 0 {
+			m.nextFileIndex = i
+			break
+		}
+		if _, ok := openedIters[i]; ok {
+			continue
+		}
+		if bytes.Compare(file.endKey, key) <= 0 {
+			continue
+		}
+		iter, err := m.openIter(file)
+		if err != nil {
+			m.err = err
+			return false
+		}
+		if iter.Seek(key) {
+			heap.Push(&m.heap, mergingIterItem{
+				Iterator: iter,
+				index:    i,
+			})
+		} else {
+			m.err = iter.Error()
+			closeErr := iter.Close()
+			m.err = goerrors.Join(m.err, closeErr)
+			if m.err != nil {
+				return false
+			}
+		}
+	}
+	// Although we have opened all files whose range contains the seek key,
+	// it is possible the seeked key is greater than the original seek key.
+	// So we need to check if the next file needs to be opened.
+	//
+	// Consider the following case:
+	//
+	// file 1: a--------f
+	// file 2:   b------f
+	// file 3:       d--f
+	//
+	// If we seek to "c", we will open file 1 and file 2 first. However, the
+	// smallest key which is greater than "c" is "e", so we need to open file 3
+	// to check if it has keys less than "e".
+	return m.maybeOpenNextFiles()
+}
+
+func (m *mergingIter) First() bool {
+	return m.Seek(nil)
+}
+
+func (m *mergingIter) Next() bool {
+	m.err = nil
+	m.curKey = append(m.curKey[:0], m.heap[0].UnsafeKey()...)
+	//revive:disable:empty-block
+	for m.next() && bytes.Equal(m.heap[0].UnsafeKey(), m.curKey) {
```

Comment:
the content of the key in those file is `row key + row id`, so we don't need to handle duplicates(should not have those) ?

Reply:
It can be any content. `DiskSorter` is a common implementation for `ExternalSorter` which will automatically skip duplicate keys. See the comment in `ExternalSorter` .